File: LeetCode/0028.go
package leetcode

func strStr(haystack string, needle string) int {
	m, n := len(haystack), len(needle)
	next := make([]int, n+1)
	getnext := func() {
		j, k := 0, -1
		next[0] = k
		for j < n-1 {
			for k != -1 && needle[j] != needle[k] {
				k = next[k]
			}
			j, k = j+1, k+1
			next[j] = k
			if needle[j] == needle[k] {
				next[j] = next[k]
			}
		}
	}
	getnext()
	i, j := 0, 0
	for i < m && j < n {
		for j != -1 && haystack[i] != needle[j] {
			j = next[j]
		}
		i++
		j++
	}
	if j >= n {
		return i - n
	}
	return -1

}

File: leetcode/0028.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int m = haystack.size(), n = needle.size();
        if (n == 0) return 0;
        vector<int> next(n + 1, 0);
        auto getnet = [&]() {
            int i = 0, k = -1;
            next[i] = k;
            while (i < n) {
                while (k != -1 && needle[i] != needle[k]) {
                    k = next[k];
                }
                next[++i] = ++k;
                if (needle[k] == needle[i]) next[i] = next[k];
            }
        };
        getnet();
        int i = 0, j = 0;
        for (; i < m && j < n;) {
            if (j == -1 || haystack[i] == needle[j]) {
                j++;
                i++;
            } else {
                j = next[j];
            }
        }
        if (j == n) {
            return i - j;
        } else {
            return -1;
        }
    }
};